sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java

```diff
@@ -80,8 +80,23 @@ public void processElement(ProcessContext ctx) {
         case INTERSECT:
           if (leftRows.iterator().hasNext() && rightRows.iterator().hasNext()) {
             if (all) {
-              for (Row leftRow : leftRows) {
-                ctx.output(leftRow);
+              Iterator<Row> iter = leftRows.iterator();
+              int leftCount = 0;
+              int rightCount = 0;
+              while (iter.hasNext()) {
+                iter.next();
+                leftCount++;
+              }
+              iter = rightRows.iterator();
+              while (iter.hasNext()) {
+                iter.next();
+                rightCount++;
+              }
+
+              // output MIN(m, n)
+              iter = (leftCount <= rightCount) ? leftRows.iterator() : rightRows.iterator();
+              while (iter.hasNext()) {
```

Comment:
Looking at this I am not sure I understand how SQL `INTERSECT` is supposed to work, can you add a brief explanation comment?

Reply:
I see. What I was missing is how this whole thing works, i.e. that it runs `CoGBK` with keys and values being the whole row, producing a `CoGBKResult` with a collection of duplicates for each distinct row, which we then emit here.

Reply:
Can we instead potentially implement it with a combiner on both inputs, that counts the rows? I.e. `row -> KV<row,count> -> CoGBK -> emit min(count) number of times`? Would it be more optimal?

Reply:
If we cannot avoid shuffle, passing around value or counts will sometime not really impact performance. It really depends. I remember one problem faced by shuffle is too many small files. Because by the nature of shuffle, there will might be m * n connections where m is the number of mappers, n is the number of keys/small files on each mapper.  Some optimization on shuffle that might be done is compact files into larger files, or pipeline might enable external shuffle service. 

Basically I think one solution is not always be optimal all the time.